### PR TITLE
homeTask7

### PR DESCRIPTION
**Pull request by VitalyZenkou**

### CODE REVIEW

File: com/course/zenkouv/src/test/java/webDriverTest/Onliner.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+    @Test
+    public void loadOnlinerPage() {
+        chromeDriver.get(URL);
+        Assert.assertEquals(chromeDriver.getTitle(), "Каталог Onliner.by");
+    }
+
+    @Test(dependsOnMethods = "loadOnlinerPage")
+    public void authorization() {
+        wait.until(webDriver -> chromeDriver.findElement(By.id(OnlinerLocators.USER_BAR_ID))).click();
```
Comment (nemec8g) [inline com/course/zenkouv/src/test/java/webDriverTest/Onliner.java]: move `wait.until(webDriver -> chromeDriver.findElement(By by)` to separate class WaitUtil

File: com/course/zenkouv/src/test/java/webDriverTest/Onliner.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        wait.until(webDriver -> chromeDriver.findElement(By.cssSelector(OnlinerLocators.LOGIN_FIELD_CSS))).sendKeys(LOGIN);
+        chromeDriver.findElement(By.cssSelector(OnlinerLocators.PASSWORD_FIELD_CSS)).sendKeys(PASSWORD);
```
Comment (nemec8g) [inline com/course/zenkouv/src/test/java/webDriverTest/Onliner.java]: you may create object `Onliner` with methods like `login` , goToCatalog , buyProduct, and operate with this object in test

File: com/course/zenkouv/src/test/java/webDriverTest/Onliner.java [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        chooseRandomAvailableProduct();
+        addProductToCart();
+        Assert.assertTrue(wait.until(webDriver -> chromeDriver.findElements(By.xpath(OnlinerLocators.CART_XPATH)).size() > 0),
+                "The product wasn't added to cart!");
+    }
+
+    @Test(dependsOnMethods = "addRandomProductToCart")
+    public void checkProductInCart() {
+        wait.until(webDriver -> chromeDriver.findElement(By.xpath(OnlinerLocators.CART_BUTTON_XPATH))).click();
+        Assert.assertTrue(wait.until(webDriver -> chromeDriver.findElements(By.className(OnlinerLocators.CART_PRODUCT_CLASS_NAME)).size() > 0));
```
Comment (nemec8g) [inline com/course/zenkouv/src/test/java/webDriverTest/Onliner.java]: Move this assertion as `IsElementPresent(By By)`  to Validator and reuse in all places

File: com/course/zenkouv/src/test/java/webDriverTest/Onliner.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    private WebElement getRandomWebElement(List<WebElement> elements, String message) {
+        if (!elements.isEmpty()) {
+            return elements.get(new Random().nextInt(elements.size()));
+        } else {
+            throw new NoSuchElementException(message);
+        }
+    }
+
+    private String getCatalogSelectorCssLocator(String dataId) {
+        return "div.catalog-navigation-list__category[data-id='" + dataId + "']";
```
Comment (nemec8g) [inline com/course/zenkouv/src/test/java/webDriverTest/Onliner.java]: use String.format

File: com/course/zenkouv/src/test/java/constant/OnlinerLocators.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+    public static final String CATALOG_NAVIGATION_CLASS_NAME = "catalog-navigation-classifier__item";
+    public static final String DATA_ID = "data-id";
+    public static final String CATALOG_NAVIGATION_ASIDE_ITEM_CLASS_NAME = "catalog-navigation-list__aside-item";
+    public static final String CATALOG_NAVIGATOR_DROP_DOWN_ITEM_CLASS_NAME = "catalog-navigation-list__dropdown-item";
+    public static final String PRODUCT_CLASS_NAME = "schema-product__group";
+    public static final String PRODUCT_TITLE_CLASS_NAME = "schema-product__title";
+    public static final String CART_BUTTON_XPATH = "//a[contains(@data-href,'https://cart.onliner.by')]";
+    public static final String PRODUCT_SIDE_BOX_CLASS_NAME = "product-aside__box";
+    public static final String PRODUCT_PRICE_CLASS_NAME = "schema-product__price";
+    public static final String CART_XPATH = "//span[contains(.,'Корзина')]";
```
Comment (nemec8g) [inline com/course/zenkouv/src/test/java/constant/OnlinerLocators.java]: bad locator  - it depends on language

File: com/course/zenkouv/src/test/java/webDriver/util/filter/OfferFilter.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import webDriver.constant.OnlinerLocators;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebElement;
+import webDriver.util.validator.ElementValidator;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class OfferFilter {
+
```
Comment (nemec8g) [inline com/course/zenkouv/src/test/java/webDriver/util/filter/OfferFilter.java]: OfferFilter and ProductFilter almost the same 
create `ElementFilter` with two param `List<WebElement>` and `By  by`

File: com/course/zenkouv/src/test/resources/testSuit.xml [hunk 7]
```diff
@@ -2,7 +2,8 @@
 <suite name="Test Suit">
     <test name="DemoTest" parallel="methods" thread-count="1">
         <classes>
-            <class name="calculator.CalculatorTest"/>
+            <class name="test.calculator.CalculatorTest"/>
```
Comment (nemec8g) [inline com/course/zenkouv/src/test/resources/testSuit.xml:5]: no need to run all tests in one scope 
create new xml for ui tests

File: com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Page.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    public void moveTo(WebElement element) {
+        Actions action = new Actions(driver);
+        action.moveToElement(element).build().perform();
+    }
+
+    public String getPageTitle() {
+        return driver.getTitle();
+    }
+
+    public WebElement findElementWithWait(By locator){
```
Comment (nemec8g) [inline com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Page.java]: use minimal accessor in all methods. i think protected will suit you

File: com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Offers.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    public Offers(WebDriver driver) {
+        super(driver);
+        PageFactory.initElements(super.driver, this);
+    }
+
+    public List<WebElement> getOffers() {
+        return offers;
+    }
+
+    public void addProductToCart() {
```
Comment (nemec8g) [inline com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Offers.java]: addRandomProductToCart

File: com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/CatalogOnliner.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+    public Auth openAuthPage() {
+        authBar.click();
+        return new Auth(super.driver);
+    }
+
+    public Category goToCategory() {
+        WebElement category = WebDriverRandomUtils.getRandomWebElement(categories, "There isn't any catalog item!");
+        category.click();
+
```
Comment (nemec8g) [inline com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/CatalogOnliner.java]: UEL

File: com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Cart.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+    public boolean isCleanedCart() {
+        if (!products.isEmpty()) {
+            products.forEach(product ->
+            {
+                product.findElement(By.className(OnlinerLocators.TRASH_CLASS_NAME)).click();
+                (new WebDriverWait(super.driver, 1)).until(ExpectedConditions.invisibilityOf(product));
+            });
+            leaveCart();
+            return true;
```
Comment (nemec8g) [inline com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Cart.java]: `  return isEmptyCart();`  in both cases?

File: com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Auth.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+    @FindBy(css = OnlinerLocators.AUTH_SUBMIT_BUTTON_CSS)
+    private WebElement submitButton;
+
+    public Auth(WebDriver driver) {
+        super(driver);
+        PageFactory.initElements(super.driver, this);
+    }
+
+    public boolean authorization(String login, String password) {
```
Comment (nemec8g) [inline com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Auth.java]: rename to login

File: com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Auth.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+        loginField.sendKeys(login);
+        passwordField.sendKeys(password);
+        submitButton.click();
+        return isElementPresent(By.className(OnlinerLocators.PROFILE_CLASS_NAME));
```
Comment (nemec8g) [inline com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Auth.java]: make assertion here and return instance of MainPage or what ever

File: com/course/zenkouv/src/test/java/test/webDriverTest/OnlinerTest.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+    @BeforeClass
+    public void setup() {
+        chromeDriver = WebDriverFactory.getChromeDriver();
+        chromeDriver.get(url);
+    }
+
+    @Test
+    public void productAddedToCart() {
+        CatalogOnliner catalog = new CatalogOnliner(chromeDriver);
```
Comment (nemec8g) [inline com/course/zenkouv/src/test/java/test/webDriverTest/OnlinerTest.java]: in PageObject philosophy you should name `Page` classes with `Page` ending
`CatalogPage` `LoginPage` etc

Comment (VitalyZenkou) [reply com/course/zenkouv/src/test/java/test/webDriverTest/OnlinerTest.java]: fix

File: com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Auth.java (hunk 12 above)
Comment (VitalyZenkou) [reply com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Auth.java]: fix

File: com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/CatalogOnliner.java (hunk 10 above)
Comment (VitalyZenkou) [reply com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/CatalogOnliner.java]: fix

File: com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Offers.java (hunk 9 above)
Comment (VitalyZenkou) [reply com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Offers.java]: fix

File: com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Page.java (hunk 8 above)
Comment (VitalyZenkou) [reply com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Page.java]: fix

File: com/course/zenkouv/src/test/java/webDriver/util/filter/OfferFilter.java (hunk 6 above)
Comment (VitalyZenkou) [reply com/course/zenkouv/src/test/java/webDriver/util/filter/OfferFilter.java]: fix

File: com/course/zenkouv/src/test/resources/testSuit.xml (hunk 7 above)
Comment (VitalyZenkou) [reply com/course/zenkouv/src/test/resources/testSuit.xml:5]: fix

File: com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Auth.java (hunk 13 above)
Comment (VitalyZenkou) [reply com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Auth.java]: fix

File: com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Cart.java (hunk 11 above)
Comment (VitalyZenkou) [reply com/course/zenkouv/src/test/java/webDriver/pageObject/onliner/Cart.java]: fix, change this method

**Pull request merged**